packages/jest-resolve-dependencies/src/index.ts

```diff
@@ -64,8 +67,31 @@ class DependencyResolver {
         }
       }
 
-      if (resolvedDependency) {
-        acc.push(resolvedDependency);
+      if (!resolvedDependency) {
+        return acc;
+      }
+
+      acc.push(resolvedDependency);
+
+      // If we resolve a dependency, then look for a mock dependency
+      // of the same name in that dependency's directory.
+      resolvedMockDependency = this._resolver.getMockModule(
+        resolvedDependency,
+        path.basename(dependency),
+      );
```

Comment:
Continuing [this](https://github.com/facebook/jest/pull/8952#discussion_r324457995) discussion here, I agree with @petevdp that `getMockModule(file, dependency)` wouldn't work as expected here (I've tried it and tests fail).
After the changes done in this PR, resolving dependencies works like this:
1. Try to resolve module from dependency. If this fails, try to get mock module from `rootFolder/__mocks__` (This is done by the `getMockModule(file, dependency)` line.
2. If step 1 returned a module, check if there is a mock in the `folderWhereTestIsLocated/__mocks__` folder and add it.
@jeysal

Reply:
Still totally lost as to which module structures the different code paths actually apply to, but I guess as long as the tests pass... I guess this is one of those pieces of code with terrible naming where nobody knows how it works anymore lol